src/Client/ClickHouseClient.cpp

```diff
@@ -0,0 +1,286 @@
+        parameters.security);
+
+    ret->setCompatibleWithClickHouse();
+    return ret;
+}
+
+size_t calculatePollInterval(const ConnectionTimeouts & timeouts)
+{
+    const auto & receive_timeout = timeouts.receive_timeout;
+    constexpr size_t default_poll_interval = 1'000'000; /// in microseconds
+    constexpr size_t min_poll_interval = 5'000; /// in microseconds
+    return std::max(min_poll_interval, std::min<size_t>(receive_timeout.totalMicroseconds(), default_poll_interval));
+}
+
+}
+
+ClickHouseClient::ClickHouseClient(ConnectionParameters params_, Poco::Logger * logger_)
+    : params(params_)
+    , connection(createConnection(params))
+    , poll_interval(calculatePollInterval(params.timeouts))
+    , logger(logger_)
+{
+}
+
+void ClickHouseClient::reset()
+{
+    cancelled = false;
+    processed_rows = 0;
+    server_exception = nullptr;
+}
+
+void ClickHouseClient::executeQuery(const String & query, const String & query_id, bool fail_quick)
+{
+    assert(!has_running_query);
+    has_running_query = true;
+
+    reset();
+
+    bool suppress_error_log {false};
+    while (true)
+    {
+        try
+        {
+            connection->sendQuery(
+                params.timeouts,
+                query,
+                {} /*query_parameters*/,
+                query_id,
+                QueryProcessingStage::Complete,
+                nullptr /*settings*/,
+                nullptr /*client_info*/,
+                false);
+
+            break;
+        }
+        catch (const Exception & e)
+        {
+            if (fail_quick)
+                e.rethrow();
+
+            /// connection lost
+            if (!connection->checkConnected())
+            {
+                if (!suppress_error_log)
+                    LOG_ERROR(logger, "Connection lost");
+                /// set the connection not connected so that sendQuery will reconnect
+                connection->disconnect();
+                std::this_thread::sleep_for(std::chrono::seconds(2));
+            }
+            /// Retry when the server said "Client should retry" and no rows has been received yet.
+            else if (processed_rows == 0 && e.code() == ErrorCodes::DEADLOCK_AVOIDED)
+            {
+                if (!suppress_error_log)
+                    LOG_ERROR(logger, "Got a transient error from the server, will retry in 1 second");
+                std::this_thread::sleep_for(std::chrono::seconds(1));
+            }
+            else
+            {
+                has_running_query = false;
+                e.rethrow();
+            }
+
+            /// Otherwise, it will keep generating the same error log again and again until the connection is back.
+            suppress_error_log = true;
+        }
+    }
+}
+
+void ClickHouseClient::executeInsertQuery(const String & query, const String & query_id)
+{
+    executeQuery(query, query_id);
+    receiveEndOfQuery();
+}
+
+std::optional<Block> ClickHouseClient::pollData()
```

Comment:
we may want to expose a timeout to `pollData(...)` to let the caller to decide how long we like to poll / wait for network packets etc. We may need have a sentinel to tell user this is the end of stream (done with the query etc) ? Having said this StorageDistributed and the related code has all of these kind of handling.

Reply:
`ClickHouseClient` takes `ConnectionParams`, which contains `ConnectionTimeouts` which have all kinds of timeout settings. That's why `pollData` itself does not need a timeout parameter.

Reply:
it could be different, those params settings is for connection, this one is for how long we wait for data which could be different ?

Reply:
Let's defer all of these for second PR since when we reuse most of the exiting StorageDistributed related code, lots of things are already considered probably.